include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h

```diff
@@ -163,16 +166,33 @@ single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __ou
             sycl::group_barrier(group);
             std::uint32_t tile_id = tile_id_lacc[0];
 
-            auto current_offset = (tile_id*elems_in_tile);
-            auto next_offset = ((tile_id+1)*elems_in_tile);
-            if (next_offset > n)
-                next_offset = n;
-            auto in_begin = __in_rng.begin() + current_offset;
-            auto in_end = __in_rng.begin() + next_offset;
-            auto out_begin = __out_rng.begin() + current_offset;
-
-            if (current_offset >= n)
+            // Global load into local
+            auto wg_current_offset = (tile_id*elems_in_tile);
+            auto wg_next_offset = ((tile_id+1)*elems_in_tile);
+            size_t wg_local_memory_size = elems_in_tile;
+            if (wg_current_offset >= n)
                 return;
+            if (wg_next_offset >= n) {
+                wg_local_memory_size = n - wg_current_offset;
+                wg_next_offset = n; // Not needed
+            }
+
+            // TODO: vectorize loads, where possible
+            if (wg_next_offset <= n) {
```

Comment:
Something wrong with the logic here. `wg_next_offset` is never greater than `n`, so the second branch will never be taken.

Reply:
Ya iv fixed this by having `wg_next_offset` not set to n if too large. This means the second branch is taken in cases where we might read memory out of the allocated bounds.